### PR TITLE
feat: allow generation of url with key for non-yet-existing content

### PR DESCRIPTION
**Pull request by dzdidi**
#  allow generation of url with key for non-yet-existing content

### CODE REVIEW

File: lib/client/index.js
Comment (mainmode1) [inline lib/client/index.js:212]: Small point, but perhaps `opts.encrypt` instead to be consistent with `put()` method.

Comment (mainmode1) [reply lib/client/index.js:212]: Then for completeness:
```diff
@@ -85,10 +85,9 @@ class Client {
    */
   async put (path, content, opts = {}) {
     const fullPath = this._fullPath(path)
+    const parsed = this._parseURL(path)
 
-    content = opts.encrypt ? await this._encrypt(fullPath, content) : content
+    content = (opts.encrypt || parsed?.encryptionKey) ? await this._encrypt(fullPath, content, parsed?.encryptionKey) : content
 
     const record = await Record.create(this._keyPair, fullPath, content, { metadata: { encrypted: opts.encrypt } })
 
@@ -554,11 +553,12 @@ class Client {
   /**
    * @param {string} path
    * @param {Uint8Array} content
+   * @param {Uint8Array} key
    */
-  async _encrypt (path, content) {
+  async _encrypt (path, content, key) {
     if (content.length === 0) return content
 
-    const encryptionKey = await this._generateEncryptionKey(path)
+    const encryptionKey = key ?? await this._generateEncryptionKey(path)
 
     // <nonce><authentication tag><encrypted content>
     const result = Buffer.alloc(sodium.crypto_secretbox_NONCEBYTES + sodium.crypto_secretbox_MACBYTES + content.length)
```

Which will allow `put()` using same uri and you'll have a r/w 'shareable' encrypted file.

Comment (dzdidi) [reply lib/client/index.js:212]: Good point. Chaning. I was on the edge due to possible confusion that returned url itself can be encrypted

Comment (mainmode1) [reply lib/client/index.js]: Actually scratch my second comment. This would be a bigger change as `Record` signatures won't match when `put`ing to remote relay without key delegation.

Comment (dzdidi) [reply lib/client/index.js]: > when puting to remote relay without key delegation.

not sure I understand what do you mean by "without key delegation." and thus why signatures will not match

Comment (mainmode1) [reply lib/client/index.js]: Let's say you try this at the end of the same test you modified:
```js
  const urlWithEncryptKey = await a.createURL('bar', { encrypt: true })
  await b.put(urlWithEncryptKey, value, { awaitRelaySync: true })
  t.alike(await a.get('bar'), value, 'get shared encrypted file (e2e)')
  // throws Error: Invalid signature
```
`b` **signs** and puts the record, then `a` attempts to verify on get using (its own) z32 id in the url as pubkey => sig mismatch.

One workaround would be to include signer `b`'s pubkey in the encoded url as query param, ie. `{ shared: [ key: <z32 id string>, expiry?: <timestamp> ] }`. When put, the `shared` obj array is saved with the record metadata, then peers, ie. `a`, could verify the signature against those pubkeys provided the expiry (if specified) has not elapsed.

However, this is not secure as it allows publishing arbitrary data to a peer's relay & id.

Easier: `a` creates a signed delegate record/blob for `b`'s pubkey + expiry beforehand (similar to [NIP26](https://github.com/nostr-protocol/nips/blob/master/26.md)).

Harder: 3-round [MuSig](https://eprint.iacr.org/2018/068).

Comment (dzdidi) [reply lib/client/index.js]: ok, it makes sense that `b` should not be able to write to `a`'s file. I've misunderstood the intention of [your comment](https://github.com/slashtags/web-relay/pull/52#discussion_r1415712319)  then. I thought it was for passing encryption key which can be created in by user in the way they want. Made me wonder why not pass it in opts then... Reverting

**Comment by dzdidi** [comment]
Not sure what is the process... Can this be merged (by me or @rbndg ) or should it wait for review from @Nuhvi 

prerequisite for https://github.com/slashtags/skunk-works/pull/1

**Review by Nuhvi — APPROVED** [review]

**Pull request merged**
